src/MOI/callbacks/MOI_user_cut.jl

```diff
@@ -0,0 +1,140 @@
+function moi_user_cut_xprs_optnode_wrapper(func::Function, model::Optimizer, callback_data::CD) where {CD<:CallbackData}
+    info = model.callback_table.moi_user_cut::Union{Tuple{CallbackInfo{CD}},Nothing}
+
+    if !isnothing(info)
+        push_callback_state!(model, CS_MOI_USER_CUT)
+
+        get_callback_solution!(model, callback_data.node_model)
+
+        # Apply stored cuts if any
+        if isempty(model.callback_cut_data.cut_ptrs) || !apply_cuts!(model, callback_data.node_model)
+            # Allow only one user cut solution per LP optimal node limiting
+            # two calls to guarantee th user has a chance to add a cut.
+            # If the user cut is loose the problem will be resolved anyway.
+            if Xpress.getintattrib(callback_data.node_model, Xpress.Lib.XPRS_CALLBACKCOUNT_OPTNODE) <= 2
+                func(callback_data)
+            end
+        end
+
+        pop_callback_state!(model)
+    end
+
+    return nothing
+end
+
+function MOI.set(model::Optimizer, ::MOI.UserCutCallback, ::Nothing)
+    info = model.callback_table.moi_user_cut::Union{Tuple{CallbackInfo{OptNodeCallbackData}},Nothing}
+
+    if !isnothing(info)
+        xprs_optnode_info, info
+
+        remove_xprs_optnode_callback!(model.inner, xprs_optnode_func)
+
+        model.callback_table.moi_user_cut = nothing
+    end
+
+    return nothing
+end
+
+function MOI.set(model::Optimizer, attr::MOI.UserCutCallback, func::Function)
+    MOI.set(model, attr, nothing)
+
+    xprs_optnode_info = add_xprs_optnode_callback!(
+        model.inner,
+        (callback_data::OptNodeCallbackData) -> moi_user_cut_xprs_optnode_wrapper(
+            func,
+            model,
+            callback_data,
+        )
+    )::CallbackInfo{OptNodeCallbackData}
+
+    model.callback_table.moi_user_cut = (xprs_optnode_info,)
+
+    return nothing
+end
+
+MOI.supports(::Optimizer, ::MOI.UserCutCallback) = true
+MOI.supports(::Optimizer, ::MOI.UserCut{CD}) where {CD<:CallbackData} = true
+
+function MOI.submit(
+    model::Optimizer,
+    submittable::MOI.UserCut{CD},
+    f::MOI.ScalarAffineFunction{T},
+    s::Union{MOI.LessThan{T},MOI.GreaterThan{T},MOI.EqualTo{T}}
+) where {T,CD<:CallbackData}
+    # It is assumed that every '<:CallbackData' has a 'node_model' field
+    node_model = submittable.callback_data.node_model::Xpress.XpressProblem
+
+    # TODO: Should we mark as submitted in the beginning of the function?
+    model.callback_cut_data.submitted = true
+
+    # Check callback state
+    let state = callback_state(model)
```

Comment:
"let" is really needed?
if so add comments explaining why

Reply:
I wanted it's scope to be restricted for clarity and isolation (let was used to avoid creating a function for this code block)

Reply:
The `let` block was turnet into a function